### ISSUE TITLE
server lagging

### ISSUE THREAD

**Issue by Vizvezdenec** [glinscott/fishtest#435]
Can't open any page apart from the main page which anyway seems to be not refreshing.

**Comment by vondele** [comment]
at the same time, when one gets a page something seems inconsistent:

Active - 1204 machines 7252 cores 0.48M nps (3504.46M total nps) 2464 games/minute

I highly doubt that right now there are that many machines connect, and this is also not compatible with the low number of games per minute.

**Comment by Vizvezdenec** [comment]
the same problem as I mentioned - most tasks get 0 workers assigned while 3-4 tasks get all the workers.

**Comment by ppigazzini** [comment]
@tomtor view https://github.com/glinscott/fishtest/issues/423#issuecomment-556249674 for the problem reported by @Vizvezdenec

**Comment by Vizvezdenec** [comment]
it's indeed a thing - my and @mstembera newly submitted tests don't get any workers for an hour.

**Comment by xoto10** [comment]
> at the same time, when one gets a page something seems inconsistent:
> 
> Active - 1204 machines 7252 cores 0.48M nps (3504.46M total nps) 2464 games/minute
> 
> I highly doubt that right now there are that many machines connect, and this is also not compatible with the low number of games per minute.

My server is listed 3 times in the list of machines, so I guess the number of machines quoted is 3x the real figure. Currently most say several minutes since last updated, so i guess no tests are running at the moment.

**Comment by crossbr** [comment]
My worker keeps timing out:

Worker started in C:\fishtest-master\worker ...

Memory: 34289192960
Worker version 70 connecting to https://tests.stockfishchess.org:443
Fetch task...
Worker version checked successfully in 0.547084s
Exception accessing host:
Traceback (most recent call last):
  File "C:\fishtest-master\worker\requests\packages\urllib3\connectionpool.py", line 393, in _make_request
    six.raise_from(e, None)
  File "<string>", line 2, in raise_from
  File "C:\fishtest-master\worker\requests\packages\urllib3\connectionpool.py", line 389, in _make_request
    httplib_response = conn.getresponse()
  File "C:\Python38\lib\http\client.py", line 1322, in getresponse
    response.begin()
  File "C:\Python38\lib\http\client.py", line 303, in begin
    version, status, reason = self._read_status()
  File "C:\Python38\lib\http\client.py", line 264, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "C:\Python38\lib\socket.py", line 669, in readinto
    return self._sock.recv_into(b)
  File "C:\Python38\lib\ssl.py", line 1241, in recv_into
    return self.read(nbytes, buffer)
  File "C:\Python38\lib\ssl.py", line 1099, in read
    return self._sslobj.read(len, buffer)
socket.timeout: The read operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\fishtest-master\worker\requests\adapters.py", line 413, in send
    resp = conn.urlopen(
  File "C:\fishtest-master\worker\requests\packages\urllib3\connectionpool.py", line 639, in urlopen
    retries = retries.increment(method, url, error=e, _pool=self,
  File "C:\fishtest-master\worker\requests\packages\urllib3\util\retry.py", line 261, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "C:\fishtest-master\worker\requests\packages\urllib3\packages\six.py", line 686, in reraise
    raise value
  File "C:\fishtest-master\worker\requests\packages\urllib3\connectionpool.py", line 592, in urlopen
    httplib_response = self._make_request(conn, method, url,
  File "C:\fishtest-master\worker\requests\packages\urllib3\connectionpool.py", line 395, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)
  File "C:\fishtest-master\worker\requests\packages\urllib3\connectionpool.py", line 315, in _raise_timeout
    raise ReadTimeoutError(self, url, "Read timed out. (read timeout=%s)" % timeout_value)
requests.packages.urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='tests.stockfishchess.org', port=443): Read timed out. (read timeout=15.0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\fishtest-master\worker\worker.py", line 108, in worker
    req = requests.post(remote + '/api/request_task', data=json.dumps(payload), headers={'Content-type': 'application/json'}, timeout=HTTP_TIMEOUT)
  File "C:\fishtest-master\worker\requests\api.py", line 110, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "C:\fishtest-master\worker\requests\api.py", line 56, in request
    return session.request(method=method, url=url, **kwargs)
  File "C:\fishtest-master\worker\requests\sessions.py", line 475, in request
    resp = self.send(prep, **send_kwargs)
  File "C:\fishtest-master\worker\requests\sessions.py", line 596, in send
    r = adapter.send(request, **kwargs)
  File "C:\fishtest-master\worker\requests\adapters.py", line 499, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='tests.stockfishchess.org', port=443): Read timed out. (read timeout=15.0)

**Comment by xoto10** [comment]
> Worker version checked successfully in 0.547084s

This seems quite slow, does anyone know if this is a normal figure?

We know from the past that STC jobs stress the server more than LTC, perhaps it will run smoother if we set some of the STC jobs to prio -1?

**Comment by Vizvezdenec** [comment]
and now it's completely dead.

**Comment by xoto10** [comment]
Looking a bit better now. We're continuing to see what else we can do to improve it, the newer version of MongoDB should be better, but might need code changes to make the most of it. On the other hand, the I/O is a bit weaker on this machine so that might make it a bit slower than before.

@Vizvezdenec You can put my job back to 3000 otherwise it will just hog lots of cores, it looks like it will be a long one. That way we can get a few other things off the queue

**Comment by Vizvezdenec** [comment]
looking a bit better but tasks not getting workers is still an issue

**Comment by xoto10** [comment]
That's just how it works, it allocates to the jobs with the highest internal_prio. Some existing jobs have a prio well behind where it should be now since the system wasn't working well (at all?) for quite a while. It's running late. So jobs started in the last several hours are actually relatively ahead, e.g. "only" 5 hours behind compared to 24 hours behind for some, say.
As the system catches up it will sort itself out. When running smoothly all the internal_prio values should be about the same, and ahead of the current time. Assuming people don't overload it with a bucketload of new tests of course, I mean I don't really understand why people were submitting new tests today when it was clearly already broken / struggling. Anyway, hopefully on the mend now ... :-)

**Comment by Vizvezdenec** [comment]
Well I'm just saying that tests submitted today for STC get 0 workers for basically 16+ hours. Whatever causes it is not smth that works well.

**Comment by xoto10** [comment]
The system was broken. not working. dead. What do you expect, all the jobs to miraculously finish even though there's no working system?

And what doesn't help is 75% of the workers are on threaded jobs ... for some reason.

**Comment by Vizvezdenec** [comment]
75% of workers were on LTCs, random STCs, whatever. Meanwhile other LTCs with the same prio got 0 workers.
I don't see system being "fully dead" right now since tasks are still ongoing and games are running. But workers distribution looks like a huge mess right now while fistest is up and running for a couple of hours already.

**Comment by ppigazzini** [comment]
@Vizvezdenec the work done in #434 should have solved the lagging problem.
The tasks/workers problem should work out by itself, let me know if this issue can be closed.

**Comment by Vizvezdenec** [comment]
There still seems to be some problems with worker assignment.
For example now with 20~ approved tasks one task is getting 1100 cores/1200, this is LTC which had it prio set from -1 to 0.

19-11-22 23:38:56 | protonspring | ps_pawnscoring2 | modify: throughput changed from 3000 to 1000
-- | -- | -- | --
19-11-22 23:38:24 | protonspring | ps_pawnscoring2 | modify: priority changed from -1 to 0


19-11-15 | pro | ps_pawnscoring2 | diff | LLR: 0.55 (-2.94,2.94) [0.00,3.50] Total: 100004 W: 16512 L: 16204 D: 67288 | sprt     @ 60+0.6 th 1cores: 1120
-- | -- | -- | -- | -- | --


While my 2 fresh STCs have 1 core for 2 tasks, Fisherman LTC gets 80 cores while having normal 3000 TP - something seem to be not right, before latest server problems I've never seen such a strange behavior. This was a thing when we had 10000 TP as a default - workers get "stuck" on the same tasks and reassigned to them till they end, but it was never a thing since we switched to default 3000.

https://i.imgur.com/dsKJq4D.png
2 fresh LTCs with 3000 TP have nothing, one LTC with 1000 TP that is 5 days old has 95% of workers - this seems to be really not right.
Maybe it has smth to do with the fact that this LTC is from "before lags" times?

**Comment by Vizvezdenec** [comment]
let me investigate this a bit more for now...

**Comment by tomtor** [comment]
When a test its priority/throughput is raised it will receive ALL new available workers until it has no longer a disadvantage compared to all other tests, which DID receive workers while the changed test was on a low priority/througput. This has always been so, it is not related to the recent performance issues.

So raising throughput/priority has a larger impact than most users expect, but it is fair if the user initially requested a lower throughput/priority than standard in order to allocate more resources to others.

In general it is better to choose a low througput for less important tests and leave it at that value. Choosing a -1 priority and changing it to 0 later results in the temporary allocation of a huge number of cores, although these tests often fail quickly, so I do not consider this a real issue.

**Comment by tomtor** [comment]
I will add this to the wiki

**Comment by Vizvezdenec** [comment]
Oh now it makes much more sence.
Although it can be quite troublesome.
For example when alayant will change multicore RT from -1 prio to 0 it will eat all fishtest for 30+0.3 8 threads task which finishes only after 40k games, so not really fast at all.

**Comment by xoto10** [comment]
> For example when alayant will change multicore RT from -1 prio to 0 it will eat all fishtest for 30+0.3 8 threads task

Yes, good example. I suggest changing this to very low tp (100 or something) and setting to prio 0 now. Once it is running smoothly and the `internal_priority | -1574468240.0` field (see test details page) matches other running jobs the tp could be increased a bit (gradually!) if required.
Otherwise try to leave it until it runs naturally when there are no jobs in queue ... which may not happen for weeks ...

I notice the number of cores allocated to each test was added to the main page not too long ago (nice!), perhaps the internal_priority could be displayed as well (e.g. on the left under the test name??) so that it's easier to see if one test has a very different value to the others?

**Comment by vondele** [comment]
I don't think the prio should be displayed on the main page... that's too detailed. Having the number of cores on the individual user page as well would be nice to have...

this behavior with prio -1 -> 0 or tp low -> tp high is old, and is indeed pretty misleading. But unless we start changing the priority calculation rule significantly, the best policy is just to keep things on the low prio/tp for the full run. I don't think we should enforce that, but as a rule of thumb that's the fair way of doing things.

**Comment by tomtor** [comment]
This issue has been solved by recent PRs

**Comment by ppigazzini** [comment]
Many thanks to @tomtor for the serie of great speedup commits!

**Comment by valera-rozuvan** [comment]
Thank you! The effort to make the server Great Again (tm) is much appreciated! :)

**Comment by ppigazzini** [comment]
And many thanks to @xoto10 for the great work with mongodb indexes!
